src/land/madis_snow2ioda.py

```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env python3
+#
+# (C) Copyright 2021-2024 EMC/NCEP/NWS/NOAA
+#
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+#
+import os
+import argparse
+import netCDF4 as nc
+import numpy as np
+from datetime import datetime
+
+import pyiodaconv.ioda_conv_engines as iconv
+from collections import defaultdict, OrderedDict
+from pyiodaconv.orddicts import DefaultOrderedDict
+
+os.environ["TZ"] = "UTC"
+
+locationKeyList = [
+    ("latitude", "float"),
+    ("longitude", "float"),
+    ("stationElevation", "float"),
+    ("dateTime", "long")
+]
+
+obsvars = {
+    'snow_depth': 'totalSnowDepth',
+}
+
+AttrData = {
+}
+
+DimDict = {
+}
+
+VarDims = {
+    'totalSnowDepth': ['Location'],
+}
+
+iso8601_string = 'seconds since 1970-01-01T00:00:00Z'
+epoch = datetime.fromisoformat(iso8601_string[14:-1])
+
+class madis(object):
+    def __init__(self, filename, mask):
+        self.filename = filename
+        self.mask = mask
+        self.varDict = defaultdict(lambda: defaultdict(dict))
+        self.outdata = defaultdict(lambda: DefaultOrderedDict(OrderedDict))
+        self.varAttrs = defaultdict(lambda: DefaultOrderedDict(OrderedDict))
+        self._read()
+
+    # Open input file and read relevant info
+    def _read(self):
+        # set up variable names for IODA
+        for iodavar in ['totalSnowDepth']:
+            self.varDict[iodavar]['valKey'] = iodavar, iconv.OvalName()
+            self.varDict[iodavar]['errKey'] = iodavar, iconv.OerrName()
+            self.varDict[iodavar]['qcKey'] = iodavar, iconv.OqcName()
+            self.varAttrs[iodavar, iconv.OvalName()]['coordinates'] = 'longitude latitude'
+            self.varAttrs[iodavar, iconv.OerrName()]['coordinates'] = 'longitude latitude'
+            self.varAttrs[iodavar, iconv.OqcName()]['coordinates'] = 'longitude latitude'
+            self.varAttrs[iodavar, iconv.OvalName()]['units'] = 'mm'
+            self.varAttrs[iodavar, iconv.OerrName()]['units'] = 'mm'
+
+        # open input file name
+        ncd = nc.Dataset(self.filename, 'r')
+        vals = ncd.variables['snowDepth'][:]
+        _FillValue = ncd.variables['snowDepth'].getncattr('_FillValue')
+
+        lats = ncd.variables['latitude'][:]
+        lons = ncd.variables['longitude'][:]
+        elvs = ncd.variables['elevation'][:]
+        qflg = ncd.variables['snowDepthQCR'][:]
```

Comment:
It looks like these are integer codes? I haven't see this before. Do we have documentation of the codes somewhere? @jiaruidong2017 Are these codes also in the sfcsno / snomad dumps?

Reply:
@ClaraDraper-NOAA , this code is very similar with the code used for processed IMS fv3 snow depth. This is python ioda converter when raw nc, csv, hd5 files are used as input. As @jiaruidong2017 talked above, all bufr data used yaml files and burfr2ioda.x to convert the bufr data into ioda format in present.

Reply:
@jiaruidong2017 Does bufr2ioda.x use these quality codes? Or, are they used in the generation of the bufr file?

Reply:
> @ClaraDraper-NOAA , this code is very similar with the code used for processed IMS fv3 snow depth. This is python ioda converter when raw nc, csv, hd5 files are used as input. As @jiaruidong2017 talked above, all bufr data used yaml files and burfr2ioda.x to convert the bufr data into ioda format in present.

This can also process hourly MADIS data. This code is processing hourly MADIS data right now.

Reply:
@ClaraDraper-NOAA No. The observation QC and Errors are defined in the yaml files.